### ISSUE TITLE
MarkFalse always overrides the "happy" status condition

### ISSUE THREAD

**Issue by antoineco** [knative/pkg#1626]
tl;dr; with the current design, whichever `MarkFalse()` happens last _in the reconciliation loop_ takes precedence over the condition that happened last _by timestamp_.

---

In the code below, the "happy" condition is overridden instantly when `MarkFalse` is called, no matter if it is currently set to a condition with a newer timestamp.

https://github.com/knative/pkg/blob/2eaf40292b3112576be1fb3cdf8b2ab2144ee645/apis/condition_set.go#L395-L404

This leads to a sub-optimal propagation of the "happy" status, because whichever `MarkFalse` comes last during the reconciliation determines what the "happy" condition will be. I would expect the order of those function calls to not matter, and the behaviour to be closer to `MarkTrue`, which ensures that the condition with the _most recent timestamp_ determines the "happy" condition (see `recomputeHappiness` func in the same file).

Question: maybe there is a good reason why things are that way. If not, does it sound reasonable to refactor `MarkFalse` and be a bit more clever about how failures propagate to "ready"?

---

Let's take an example:

```yaml
  conditions:
  - lastTransitionTime: "2020-08-13T01:49:43Z"     # <------------- recent
    message: 'The adapter Service is unavailable; Configuration "zendesksource-sample"
      does not have any ready Revision.; Revision "zendesksource-sample-pp6k6"
      failed with message: secret "zendesksource" not found.'
    reason: MissingSecret
    status: "False"
    type: Deployed
  - lastTransitionTime: "2020-08-13T01:47:39Z"
    message: The receive adapter did not report its public URL yet
    reason: FailedSync
    status: "False"
    type: Ready
  - lastTransitionTime: "2020-08-13T01:46:17Z"
    status: "True"
    type: SinkProvided
  - lastTransitionTime: "2020-08-13T01:47:39Z"     # <------------- older
    message: The receive adapter did not report its public URL yet
    reason: FailedSync
    status: "False"
    type: TargetSynced
```

Here, the `Deployed` condition is _newer_ than the `TargetSynced` condition, so I'm expecting the former to propagate to the `Ready` condition, and the reason to become `MissingSecret`.

It's not the case because in practice `TargetSynced` is always False when the adapter is down and, because of the way `MarkFalse` is implemented, my "happy" condition is _always_ set to `FailedSync` because it's evaluated last in the reconciliation logic.

---

/area API
/kind question

**Comment by mattmoor** [comment]
cc @n3wscott 

Feels largely informational (vs. semantic), but I'd guess we'll need to stage things as it may break downstream unit testing.

@dprotaso is out, but FYI.

This may also cause some ripples for Tekton, cc @ImJasonH @vdemeester @afrittoli

**Comment by n3wscott** [comment]
The contract has always been "last set wins" for false beats unknown beats true.

You could add logic to your mark method to not call down to the manager if the condition is already set?

It is not clear how the api could change to allow this to work without assuming what the caller intends.

**Comment by antoineco** [comment]
I can understand "false beats true", but I'm having a hard time understanding why `MarkTrue` would propagate the _newest by timestamp_ to "happy" while `MarkFalse` does something different. My expectation was that I wouldn't have to compare conditions myself to always ensure "happy" reflects _latest False_.
